File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/util/IncludesUtil.java
package net.ravendb.abstractions.util;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import net.ravendb.abstractions.basic.Tuple;
import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.closure.Action2;
import net.ravendb.abstractions.json.JTokenExtensions;
import net.ravendb.abstractions.json.linq.RavenJArray;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;

import org.apache.commons.lang.StringUtils;


public class IncludesUtil {

  private final static Pattern INCLUDE_PREFIX_REGEX = Pattern.compile("[^(]*(\\([^(]+\\))");

  private static IncludePath getIncludePath(String include) {
    IncludePath result = new IncludePath();
    result.setPath(include);
    Matcher matcher = INCLUDE_PREFIX_REGEX.matcher(include);
    if (matcher.matches() && matcher.groupCount() >= 1) {
      result.setPrefix(matcher.group(1));
      result.setPath(result.getPath().replace(result.getPrefix(), ""));
      result.setPrefix(result.getPrefix().substring(1, result.getPrefix().length() - 1));
    }
    return result;
  }


  private static void executeInternal(RavenJToken token, String prefix, Action2<String, String> loadId) {
    if (token == null) {
      return; // nothing to do
    }

    switch(token.getType()) {
    case ARRAY:
      for (RavenJToken item : (RavenJArray) token) {
        executeInternal(item, prefix, loadId);
      }
      break;
    case STRING:
      loadId.apply(token.value(String.class), prefix);
      break;
    case INTEGER:
      loadId.apply(token.value(Long.class).toString(), prefix);
      break;
      // here we ignore everything else
      // if it ain't a string or array, it is invalid
      // as an id
    }
  }

  private static class IncludePath  {
    private String path;
    private String prefix;

    public String getPath() {
      return path;
    }
    public void setPath(String path) {
      this.path = path;
    }
    public String getPrefix() {
      return prefix;
    }
    public void setPrefix(String prefix) {
      this.prefix = prefix;
    }

  }

  public static void include(RavenJObject document, String include, final Action1<String> loadId) {
    if (StringUtils.isEmpty(include) || document == null) {
      return ;
    }

    IncludePath path = getIncludePath(include);

    for (Tuple<RavenJToken, RavenJToken> token : JTokenExtensions.selectTokenWithRavenSyntaxReturningFlatStructure(document, path.getPath())) {
      executeInternal(token.getItem1(), path.getPrefix(), new Action2<String, String>() {

        @Override
        public void apply(String value, String prefix) {
          value = (prefix != null ? prefix + value : value);
          loadId.apply(value);
        }
      });
    }
  }
}

File: src/Raven.Client/Util/IncludesUtil.cs
using System;
using System.Globalization;
using System.Text.RegularExpressions;
using Raven.Abstractions.Json;
using Raven.Imports.Newtonsoft.Json.Linq;
using Raven.Json.Linq;

namespace Raven.Abstractions.Util
{
    public class IncludesUtil
    {
        private readonly static Regex IncludePrefixRegex = new Regex(@"(\([^\)]+\))$",
            RegexOptions.Compiled |
            RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
        private readonly static Regex IncludeSuffixRegex = new Regex(@"(\[[\{0\}\/][^\]]+\])$",
         RegexOptions.Compiled |
         RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

        


        private static void ExecuteInternal(RavenJToken token, string addition, Func<string, string, bool> loadId)
        {
            if (token == null)
                return; // nothing to do

            switch (token.Type)
            {
                case JTokenType.Array:
                    foreach (var item in (RavenJArray)token)
                    {
                        ExecuteInternal(item, addition, loadId);
                    }
                    break;
                case JTokenType.String:
                    var value = token.Value<string>();
                    // we need to check on both of them, with id & without id
                    // because people will do products/1 and detaisl/products/1 and want to be able
                    // to include on that
                    loadId(value, addition);
                    if (addition != null)
                        loadId(value, null);
                    break;
                case JTokenType.Integer:
                    try
                    {
                        loadId(token.Value<long>().ToString(CultureInfo.InvariantCulture), addition);
                    }
                    catch (OverflowException)
                    {
                        loadId(token.Value<ulong>().ToString(CultureInfo.InvariantCulture), addition);
                    }
                    break;
                    // here we ignore everything else
                    // if it ain't a string or array, it is invalid
                    // as an id
            }
        }

        private class IncludePath
        {
            public string Path;
            public string Addition;
        }

        private static IncludePath GetIncludePath(string include, out bool isPrefix)
        {
            isPrefix = false;
            var result = new IncludePath { Path = include };
            var match = Match.Empty;
            var matchPrefix = IncludePrefixRegex.Match(include);
            if (matchPrefix.Success)
            {
                match = matchPrefix;
                isPrefix = true;
            }
            else
            {
                var matchSuffix = IncludeSuffixRegex.Match(include);
                if (matchSuffix.Success)
                    match = matchSuffix;
            }

            if (match.Success && match.Groups.Count >= 2)
            {
                result.Addition = match.Groups[1].Value;
                result.Path = result.Path.Replace(result.Addition, "");
                result.Addition = result.Addition.Substring(1, result.Addition.Length - 2);
            }
            return result;
        }

        public static void Include(RavenJObject document, string include, Func<string, bool> loadId)
        {
            if (string.IsNullOrEmpty(include) || document == null)
                return;
            bool isPrefix;
            var path = GetIncludePath(include, out isPrefix);

            foreach (var token in document.SelectTokenWithRavenSyntaxReturningFlatStructure(path.Path))
            {
                ExecuteInternal(token.Item1, path.Addition, (value, addition) =>
                {
                    value = (addition != null ?
                    (isPrefix ? addition + value : string.Format(addition, value)) : value);
                    return loadId(value);
                });
            }
        }
    }
}
